src/hubspot_client.js

```diff
@@ -0,0 +1,629 @@
+import _ from 'underscore';
+
+const hubspot = require('@hubspot/api-client');
+
+// If customer does not have name anywhere, this is used as placeholder when creating user
+export const MISSING_NAME_PLACEHOLDER = '[UNNAMED_CUSTOMER]';
+
+// How many times should the client attempt to retry the API call if it fails.
+const API_CALL_RETRIES = 5;
+
+// TODO: Handle errors
+
+// When processing user from our database, we do not want to send everything to hubspot
+// This schema servers as a way to check types of data before sending them to hubspot
+// and also as a filter which only allows the specified fields to be in the final
+// object before it's sent to hubspot
+const ALLOWED_USER_FIELDS = {
+    apify_id__c: _.isString,
+    annualrevenue: _.isNumber,
+
+    email: _.isString,
+    hs_additional_emails: _.isString,
+
+    subscription_plan: _.isString,
+    subscription_price: _.isNumber,
+    segment_paying_user: _.isString,
+
+    firstname: _.isString,
+    lastname: _.isString,
+
+    company: _.isString,
+    address: _.isString,
+    city: _.isString,
+    state: _.isString,
+    zip: _.isString,
+    country: _.isString,
+    eu_vat_number__c: _.isString,
+    billing_email__c: _.isString,
+
+    lifecyclestage: _.isString,
+};
+
+// Same as above, but for invoice
+const ALLOWED_INVOICE_FIELDS = {
+    apify_id__c: _.isString,
+    apify_user_id: _.isString,
+    invoice_number: _.isString,
+    invoice_type: _.isString,
+    issue_date: _.isDate,
+    price: _.isNumber,
+    currency: _.isString,
+    is_draft: _.isString,
+    payment_due_days: _.isNumber,
+    payment_status: _.isString,
+    paid_date: _.isDate,
+    invoice_description: _.isString,
+    price_before_tax: _.isNumber,
+    price_usd: _.isNumber,
+    price_before_tax_usd: _.isNumber,
+    tax_country_code: _.isString,
+    cancel_date: _.isDate,
+};
+
+// TODO: The helper function can be replaced with JSON schema validator (AJV)
+// but then we need another function which replaces dates with their stringified versions
+
+/**
+ * This helper function cleans our database objects of unnecessary fields and also compares data types to prevent
+ * errors. Schemas which are checked against the objects are above. If the key in schema is object this function
+ * does recursive check.
+ * @param {Object} data Object from our database with all values
+ * @param {Object} allowedValues Allowed keys and their datatypes
+ * @return {Object} Object with only keys allowed in the schema
+ */
+export function cleanAndCompareWithSchema(data, allowedValues) {
+    const cleanedData = {};
+    Object.keys(allowedValues).forEach((key) => {
+        // Skip schema fields that are not present in data
+        if (typeof data[key] === 'undefined') return;
+        // Save values set to null without typechecking
+        if (data[key] === null) {
+            // cleanedData[key] = null;
```

Comment:
This is commented out on purpose? :)

Reply:
Actualy this is copyied from current salesforce client without any changes. Reason is probably that salesforce did not accept null values, looking at hubspot it does not either, so I guess we can change this to empty string or something if it causes issues.